src/ZkFold/Base/Algebra/Polynomials/Univariate.hs

```diff
@@ -1,54 +1,177 @@
 
 instance (Ring c, Eq c) => AdditiveGroup (Poly c) where
-    negate (P cs) = P $ map negate cs
-
-instance (Ring c, Eq c) => MultiplicativeSemigroup (Poly c) where
-    P l * P r = removeZeros $ P $ go l r
+    negate (P cs) = P $ fmap negate cs
+
+instance (Field c, Eq c) => MultiplicativeSemigroup (Poly c) where
+    -- | If it is possible to calculate a primitive root of unity in the field, proceed with FFT multiplication.
+    -- Otherwise default to Karatsuba multiplication for polynomials of degree higher than 64 or use naive multiplication otherwise.
+    -- 64 is a threshold determined by benchmarking.
+    P l * P r
+      | V.null l = P V.empty
+      | V.null r = P V.empty
+      | otherwise = removeZeros $ P $
+          case (maybeW2n, len <= 64) of
+            (Nothing, False) -> mulKaratsuba lPaddedKaratsuba rPaddedKaratsuba
+            (Nothing, True)  -> mulVector l r
+            (Just w2n, _)    -> mulDft (p + 1) w2n lPaddedDft rPaddedDft
         where
-            go [] _      = []
-            go (x:xs) ys = zipWithDefault (+) zero zero (map (x *) ys) (zero : go xs ys)
+            len :: Int
+            len = max (V.length l) (V.length r)
+
+            p :: Integer
+            p   = (ceiling @Double $ logBase 2 (fromIntegral len))
+
+            padKaratsuba :: Int
+            padKaratsuba = 2 P.^ p
+
+            padDft :: Int
+            padDft = 2 * padKaratsuba
+
+            lPaddedKaratsuba, rPaddedKaratsuba :: V.Vector c
+            lPaddedKaratsuba = l V.++ V.replicate (padKaratsuba - V.length l) zero
+            rPaddedKaratsuba = r V.++ V.replicate (padKaratsuba - V.length r) zero
+
+            lPaddedDft, rPaddedDft :: V.Vector c
+            lPaddedDft = l V.++ V.replicate (padDft - V.length l) zero
+            rPaddedDft = r V.++ V.replicate (padDft - V.length r) zero
+
+            maybeW2n :: Maybe c
+            maybeW2n = rootOfUnity $ fromIntegral (p + 1)
+
+mulDft :: forall c . (Field c, Eq c) => Integer -> c -> V.Vector c -> V.Vector c -> V.Vector c
+mulDft p w2n lPadded rPadded = c
+  where
+    pad :: Int
+    pad = 2 P.^ p
+
+    w2nInv :: c
+    w2nInv = one / w2n
+
+    nInv :: c
+    nInv = one / fromConstant (fromIntegral @_ @Natural pad)
+
+    v1Image, v2Image :: V.Vector c
+    v1Image = genericDft p w2n lPadded
+    v2Image = genericDft p w2n rPadded
+
+    cImage :: V.Vector c
+    cImage = V.zipWith (*) v1Image v2Image
+
+    c :: V.Vector c
+    c = fmap (* nInv) $ genericDft p w2nInv cImage
+
+mulKaratsuba :: forall a. Field a => V.Vector a -> V.Vector a -> V.Vector a
+mulKaratsuba v1 v2
+  | len == 1 = V.zipWith (*) v1 v2
+  | otherwise = result
+  where
+    len :: Int
+    len = V.length v1
 
-instance (Ring c, Eq c) => MultiplicativeMonoid (Poly c) where
-    one = P [one]
+    n :: Int
+    n = len `div` 2
+
+    a, b, c, d :: V.Vector a
+    (b, a) = V.splitAt n v1
+
+    (d, c) = V.splitAt n v2
+
+    ac, bd :: V.Vector a
+    ac = mulKaratsuba a c
```

Comment:
For subvectors of smaller length, we could try and use FFT, so I would call back to `(*)` instead of explicit recursion

Reply:
Nice point